sharing_portal/views.py

```diff
@@ -360,55 +328,83 @@ def share_artifact(request, artifact):
-                    reverse("sharing_portal:share", args=[artifact.pk])
+                    reverse("sharing_portal:share", args=[artifact["uuid"]])
+                )
+
+            portal_project = Project.objects.get(
+                charge_code=form.cleaned_data["project"]
+            )
+            # If the user is a member of this project
+            if any(
+                [
+                    user
+                    for user in get_project_members(portal_project)
+                    if user.username == request.user.username
+                ]
+            ):
+                trovi.set_linked_project(
+                    artifact,
+                    form.cleaned_data["project"],
+                    request.session.get("trovi_token"),
                 )
 
-            if artifact.is_reproducible and not artifact.reproducibility_project:
-                supplemental_project = create_supplemental_project(request, artifact)
-                artifact.reproducibility_project = supplemental_project
-            artifact.save()
-            if (_sync_share_targets(artifact, project_shares=form.cleaned_data['projects'])):
-                messages.add_message(request, messages.SUCCESS,
-                    'Successfully updated sharing settings')
-
-            if (z_form.is_valid() and
-                _request_artifact_dois(artifact, request_forms=z_form.cleaned_data)):
-                messages.add_message(request, messages.SUCCESS,
-                    ('Requested DOI(s) for artifact versions. The process '
-                     'of issuing DOIs may take a few minutes.'))
-
-            return HttpResponseRedirect(reverse('sharing_portal:detail', args=[artifact.pk]))
+            if is_reproducible:
+                create_supplemental_project_if_needed(request, artifact, portal_project)
+
+            if patches:
+                try:
+                    trovi.patch_artifact(
+                        request.session.get("trovi_token"), artifact["uuid"], patches
+                    )
+                    messages.add_message(
+                        request,
+                        messages.SUCCESS,
+                        "Successfully updated sharing settings.",
+                    )
+                except trovi.TroviException:
+                    messages.add_message(
+                        request,
+                        messages.ERROR,
+                        "Error updating artifact in Trovi, please try again",
+                    )
+
+            return HttpResponseRedirect(
+                reverse("sharing_portal:detail", args=[artifact["uuid"]])
+            )
     else:
+        # Use the first linked chameleon project
+        chameleon_projects = [
+            lp
+            for lp in artifact["linked_projects"]
+            if lp.split(":", 4)[3] == "chameleon"
+        ]
+        project = None
+        if chameleon_projects:
+            project = chameleon_projects[0].split(":", 4)[4]
```

Comment:
Probably would be good to make URN extraction a helper in some utils function or something. I noticed the zenodo extraction part assumed the urn would have only 3 parts

Reply:
This is a good point. If we're only trying to access the NSS, it might be good to index with -1 instead.